### PR TITLE
Log a heapdump file on OOM by default, for YARN containers.

### PR DESCRIPTION
**Pull request by anwar6953**
The result:
image

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
Perhaps we should make this change where we specify cdap-site configs? ie., coopr?

File: cdap-common/src/main/resources/cdap-default.xml [hunk 1]
```diff
@@ -356,9 +356,9 @@
 
   <property>
     <name>app.program.jvm.opts</name>
-    <value>-XX:MaxPermSize=128M ${twill.jvm.gc.opts}</value>
+    <value>-XX:MaxPermSize=128M -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=&lt;LOG_DIR&gt; ${twill.jvm.gc.opts} </value>
     <description>
-      Java options for all program containers
+      Java options for all Apache Twill containers
```
Comment (anwar6953) [inline cdap-common/src/main/resources/cdap-default.xml]: TODO: Update all the copies (in docs).

**Comment by anwar6953** [comment]
I think it makes sense to make the change here (cdap-default.xml), since "There is no overhead in running with this option, and so it can be useful for production systems where OutOfMemoryError takes a long time to surface."
- https://docs.oracle.com/javase/7/docs/webnotes/tsg/TSG-VM/html/clopts.html

**Comment by poornachandra** [comment]
This may lead to excessive disk space usage when large number of containers go OOM on a prod cluster. Since the heap dumps are written to the logs dir, they will not be cleaned up immediately. It is safer to enable this setting in a controlled manner on a case by case basis.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 2]
```diff
@@ -356,9 +356,9 @@
 
   <property>
     <name>app.program.jvm.opts</name>
-    <value>-XX:MaxPermSize=128M ${twill.jvm.gc.opts}</value>
+    <value>-XX:MaxPermSize=128M -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=&lt;LOG_DIR&gt; ${twill.jvm.gc.opts}</value>
```
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml:359]: Better separate out as a different property, something like `app.program.container.oom.opts` and use `${app.program.container.oom.opts}` here. It's easy to turn it  on/off in that way.

**Comment by anwar6953** [comment]
The original motivation was not for internal (Cask) clusters, but for end-users, so I won't make the change to coopr templates.
Will close, since the majority opinion is that we should not make this change by default.

**Pull request closed without merging**